### PR TITLE
Implemented additional caching features for #530

### PR DESCRIPTION
**Pull request by byko3y**
Parsing and application of no-cache and private directives with arguments.
Additional validation of Cache-Control header.
New cache_control_ignore, cache_resp_hdr_del configuration options.
Fixed corruption of cached responses with empty header value.

Fixed all the comments from the https://github.com/tempesta-tech/tempesta/pull/1539 here.

### CODE REVIEW

**Comment by byko3y** [comment]
The original version of the pull request is not compilable because of excessive curly brackets. Somehow git brought my working copy 1 commit before the final version so I verified the final version itself. Good news is the last commit had only cosmetic changes so it's not a big deal.

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:2339]: This is just not addressed comment https://github.com/tempesta-tech/tempesta/pull/1539/files#r775029875 . Please review the review notes once more. In short: you shouldn't fit new logic into existing states, most likely you need several new states.Introducing new states hurts, but not so much as introducing a lot of complexity with many if's and extra variables.

Comment (krizhanovsky) [inline fw/http_parser.c:2339]: This is just not addressed comment https://github.com/tempesta-tech/tempesta/pull/1539/files#r775029875 . Please review the review notes once more. In short: you shouldn't fit new logic into existing states, most likely you need several new states.Introducing new states hurts, but not so much as introducing a lot of complexity with many if's and extra variables.

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:2388]: What is a "common flag"? The comment is very confusing in context of the state name `Req_I_CC_Flag`, where "flag" is about the parser flag, see https://github.com/tempesta-tech/tempesta/pull/1539/files#r774807540

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:2514]: A comment is required - why the success of the whole Cache-Control depends on cache_control.filled ? Actually, I believe the variable is just a variant of push down just to avoid new states, which makes the code overcomplicated and probably slower because we have extremely fast FSM technology

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8914]: We should move `"` parsing into a separate state

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8901]:
```suggestion
		/* Header name comparison requires a trailing colon */
```

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8903]: Do not make a separate call with the generic allocation and TfwStr logic. You know precisely on the `__str_array_add_token()` call above that you finished to parse a header name and you can call the function with a flag, which will add the required `:`

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8843]: Please make a comment for the state referencing the RFC chapters requiring header names parsing on `no-cache` and `private`

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8988]: Please remove the comment as well

**Review by krizhanovsky — CHANGES_REQUESTED** [review]
There are still several bugs and more work is required to for the FSM workflow simplification (basically already commented in https://github.com/tempesta-tech/tempesta/pull/1539). The FSM issues are somewhat linked with https://github.com/tempesta-tech/tempesta/pull/1538 , so please talk with @Dmitry-Gouriev - we already had conversations about automata constructions.

Also #530 requires several functional tests tests including https://github.com/tempesta-tech/tempesta/issues/810#issuecomment-966690711 (note the `cache_bypass` and `cache_fullfil` configuration options, which can not be tested with unit tests) and Wiki update.

File: fw/http_parser.c
Comment (Dmitry-Gouriev) [reply fw/http_parser.c:816]: Stange rename hdr -> hbh_hdr.

File: fw/http_parser.c
Comment (Dmitry-Gouriev) [inline fw/http_parser.c:2379]: Hello! I wonder a little that the conditions `!parser->cache_control.dir_allowed` is tested at several states.
It is set to `true` at `Req_I_CC_start`. It can become `false` during parsing? Or the parser can enter these states avoiding `Req_I_CC_start` and with `false` value for condition? Believing to the given name `dir_allowed`, the check probably better be outside this function at all?

File: fw/http_parser.c
Comment (Dmitry-Gouriev) [reply fw/http_parser.c:2496]: Hello! It seems I do not understand a conception of `dir_allowed` flag (or can hardly agree with it).

The flag is set to `true` in two states `Req_I_CC_start` and `Resp_I_EoT`, reset to `false` in several other states and is checked in many other states.

It seems that the flag is used to disable (make disallowed) a subset of states.

Is not it better to provide the same by hardcoding transitions between states with `__FSM_JPM`, `__FSM_I_MOVE_xxxx`?

We can control state transition hardly and `__FSM_START` is to restore that state on the next call of the function rather than to enter a random unpredictable state.

Comment (Dmitry-Gouriev) [reply fw/http_parser.c:2496]: Looking from more general position, I'd use states to encode parser positions inside text being parsed and flags and variables to store information which was detected in the past and needed in the future. However performance consideration could dictate something different.

File: fw/hpack.c
Comment (krizhanovsky) [reply fw/hpack.c:1900]: This is for retrieving headers with empty values from the cache, right? In other words I believe this is the part of https://github.com/tempesta-tech/tempesta/pull/1551 .

Please move the fix to #1551 AND write a better comment:
```
/*
 * The header has no value, so we are going to skip the HPACK_STATE_VALUE_TEXT,
 * but still need to write CRLF at the end of the header line for a correct HTTP/1
 */
```

File: fw/hpack.c
Comment (krizhanovsky) [inline fw/hpack.c:1891]: It still makes sense to leave the condition with `unlikely`

File: fw/vhost.h
Comment (krizhanovsky) [reply fw/vhost.h]: the comment is still not addressed

File: fw/http_parser.h
Comment (Dmitry-Gouriev) [reply fw/http_parser.h:131]: I finally do not fully understand an idea of the `dir_flag `.
It seems to be unneccessary in request parser (you can make additions (`|=`) directly to `req->cache_ctl.flags` instead of `dir_flag `.
It used in 2 conditions in reqponse parser:
```C
Line 8750: 			if (parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE)
Line 8787: 		tokens = parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE ?
```
This is because the states `Resp_I_Flag_Maybe_Qal` and `Resp_I_CC_Flag_Argument_Token` are shared between several Cache-Control directives. You can divide these states into several ones to aviod sharing. If you do not like new states... Anyway, it seems, it would be simplier to use the `dir_flag` **only** to cache the directive flag for these checks, not to set `resp->cache_ctl.flags`.

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:2411]: +1 for @Dmitry-Gouriev 's comments. `dir_flag` is just a replacement for old `_acc` and I'm good with it. But Dmitry pointed out that probably we can directly set the flags, w/o collecting them in `_acc`. `_acc` stands for "accumulator" and the primary target for the field is to accumulate multi-digit number values, which seem not necessary for flags.

`cache_control.filled` still looks unnatural, but there isn't so many code around it and now FSM looks much more consistent (so I closed #1539). However, probably we can determine whether Cache-Control had a directive with introducing only one new state. In this case we won't need `filled`.

If we can get rid of both the `filled` and `dir_flag`, then we can get rid of `Req_I_CC_start` and, with one new state instead of `filled`, we'll have the same number of states and more straightforward and likely faster FSM.

BTW since `Req_I_CC_start` is executed only once on very first FSM entrance, we can initialize the variable just at the beginning of the function and we don't need a state for it.

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:6133]: we have no CRLFs in h2

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8760]: an empty line is wished here

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8753]: Ok, I see the usage of `dir_flag` versus direct update in `cache_ctl.flags`. Otherwise we'll need at least one more state them. However, neither of the *_QUAL flags are actually used besides the tests, so they should be just removed (and the tests reworked to test the things in a different way).

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8779]: Usually, we don't use else in such cases, just
```C
if (c != '"')
    __FSM_EXIT(TFW_BLOCK);
__FSM_I_MOVE_fixup(Resp_I_CC_Flag_Argument_Token, 1, 0);
```

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8807]: Looks much better now: shorter and more efficient!

File: fw/http_parser.c
Comment (krizhanovsky) [reply fw/http_parser.c:8818]: It's still worth to move `"` parsing to a separate small state

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8719]: Don't you need to reinit `dir_flag` as well?

**Review by krizhanovsky — CHANGES_REQUESTED** [review]
The FSM is much better now and we're almost "there".

Functional tests and wiki are still in TODO. It'd better to make the tests and make sure that they do pass before the final review and merge of the code.

File: fw/vhost.h
Comment (byko3y) [reply fw/vhost.h]: It is addressed: TfwCaPolicyTokens no longer exists, instead there is "TfwCaToken *capo_hdr_del".

File: fw/http_parser.c
Comment (byko3y) [reply fw/http_parser.c:8818]: Sorry, I don't understand what part of the algorithm should be separated into the new state. To me it already looks like the most simple character probing possible.

File: fw/http_parser.c
Comment (byko3y) [reply fw/http_parser.c:2424]: Agreed, RFC seems to forbid empty tokens, I'll fix that.

File: fw/http_parser.h
Comment (byko3y) [reply fw/http_parser.h:131]: Actually there is another condition in the Req_I_CC_Flag/Resp_I_Flag, so we cannot set the req->cache_ctl.flags immediately. However, I agree it's not needed anywhere else and Resp_I_Flag_Maybe_Qal should be segregated into two states.

Comment (Dmitry-Gouriev) [reply fw/http_parser.h:131]: @byko3y 
If you mean
```
WARN_ON_ONCE(!parser->cache_control.dir_flag);
```
this is about internal check for algorithm consistency rather than to parse the text.
However this consistence condition could be proved staticaly.
Is there any way to reach these states without `reg_flag` set?

File: fw/cache.c
Comment (byko3y) [reply fw/cache.c:1735]: The nesting problem mostly comes from large tab width and small line width limit. However, in this particular case the nested block is small and fits well into 80 character limit.

File: fw/http_parser.h
Comment (byko3y) [reply fw/http_parser.h:131]: Everything can be proven statically, it's more about how complex the proof is. In this case I do agree the proof is quite simple so there's not much use for this particular statement.

File: fw/vhost.h
Comment (krizhanovsky) [reply fw/vhost.h]: I meant that why not to combine all caching policy stuff, including the new configuration directives and `cc_ignore`, into TfwCaPolicy. In this sense there is no difference between TfwCaPolicyTokens and TfwCaToken.

However, I see that `capo` and `capo_hdr_del` are arrays of different sizes, not single structures, which can be combined. So we're good with the current design.

File: fw/http_parser.c
Comment (krizhanovsky) [reply fw/http_parser.c:8808]: Yes, the state is simple, and good enough. I was confused with that it allows e.g. "Cache-Control: public, no-cache="Set-Cookie,"` (`"` just after `,`), but I believe that's OK.

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:5920]: yeah, it seems we had the issue in the parsing function because it didn't process WS at the beginning. However, it seems we can move the check and transition to `Req_I_CC`

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8739]:
```suggestion
			 * Resp_I_CC_Dir_Opening_Quote => Resp_I_CC_Dir_Arg_Token =>
```

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8752]: This state is just the same as `Resp_I_Flag_No_Cache`, so the states should be unified

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8829]: This state is to allow `no-cache="Set-Cookie, "` - is this intentional? If yes, then comment is required. Otherwise, we can just omit the check and block `"` on the default transition to `TFW_BLOCK`

**Review by krizhanovsky — CHANGES_REQUESTED** [review]
There are still things to fix and clanup

File: fw/http_parser.c
Comment (byko3y) [reply fw/http_parser.c:8665]: I did some brief benchmarks on this - after every unnecessary branch removed the `__resp_parse_cache_control` function became 0.8% faster. During `TEST_RUN(http_parser, fills_hdr_tbl_for_resp)` the `__resp_parse_cache_control` takes 0.1% of the run time (thus parser is 0.008% faster with modifications), and `TEST_RUN(http_parser, cache_control)`, which feeds nothing but "cache-control" headers to the parser, leads to `__resp_parse_cache_control` working 25% of time (parser 0.2% faster, probably same increase for `__req_parse_cache_control`).
I just don't feel like performance is an issue here. Readability and robustness are though.
Also, trailing pure conditions with same statements in branches are optimized into a single unconditional statement even at "-Og" optimization level.

Comment (krizhanovsky) [reply fw/http_parser.c:8665]: First of all, there are many questions about performance: how many times you ran the test? What the test is? Did you pick the *best* times for each of the tests? Did you test different hardware?

There is still no reason to have more code if you can have less code. More code tends to be harder to read and understand, especially if the code does senseless checks.

**Comment by byko3y** [comment]
- Removed unnecessary conditions and reworked leading comma handling in cache-control parsing functions;
- Additional tests for commas in cache-control directives and their arguments;
- Reworked tfw_str_array_append_chunk();
- Reworked __tfw_strcmpspn() to match strings even if one of them lacks a trailing separator;
- Additional tests for both tfw_str_array_append_chunk and __tfw_strcmpspn().

Will update the copyright years once the final list of files is clear.

File: fw/http_parser.c
Comment (krizhanovsky) [reply fw/http_parser.c:6116]: The comment above isn't addressed. Why do we need the check and why it's not enough just to check for token at the below?

File: fw/str.c
Comment (krizhanovsky) [inline fw/str.c:721]: Please update the comment after the simplification. When the function has signature `tfw_str_array_append_chunk(TfwPool *pool, TfwStr *array, char *data, unsigned long len)` that won't be obvious how it's different from `tfw_str_add_compound()`/`__str_grow_tree()`, so please address this question (why do we need the new function) in the comment.

**Review by krizhanovsky — CHANGES_REQUESTED** [review]
Unfortunately at least one more change and one more review is required. The main issue is still overcomplecated `tfw_str_array_append_chunk()`, which must be reworked. Several small cleanups are also required.

Please make the changes in a new commit, just like you did this time.

**Comment by byko3y** [comment]
Hopefully resolved all the comments.

File: fw/str.c
Comment (krizhanovsky) [inline fw/str.c:741]:
```suggestion
	if (unlikely(!len)) {
```

File: fw/str.c
Comment (krizhanovsky) [inline fw/str.c:808]: I think it makes sense to replace
```
We don't care about array->len, but
```
with
```
array->len has no sense since this is an array of independent strings. However,
```

File: fw/str.c
Comment (krizhanovsky) [inline fw/str.c:804]: `added_len` is always zero here, so just use `len` at the below and remove `added_len` completely

File: fw/http_parser.c
Comment (krizhanovsky) [inline fw/http_parser.c:8783]: Please start the sentence form Captital letter and end it with dot.

Also please add following comment here: "We do not set TfwStr->skb for tokens, since the tokens are used in read-only fashion to compare strings and are never rewriten (TfwStr->skb is used for rewriting raw data only)."

**Review by krizhanovsky — APPROVED** [review]
Good to merge after the small cleanups

**Pull request merged**
